general/bundler_imports.md

```diff
@@ -0,0 +1,160 @@
+
+### Different strategies for npm bundler imports versioning
+
+As said in the previous part, whenever you use imports you have two places to specify versions numbers:
+
+1. The `package.json` file
+2. The `.npmbundlerrc` file
+
+Let's see what happens when the two match or differ...
+
+1. **Both version constraints in `package.json` and `.npmbundlerrc` are the same**: this is the most intuitive and logical case, where you use the same version for building and running. The only reason why you wouldn't want to do this is if the lifecycle of the provider is decoupled from your project's lifecycle, as is the usual case in DXP (because `frontend-js-react-web$react` is managed by the Infrastructure team and your project is managed by you).
+
+2. **Different version constraints in `package.json` and `.npmbundlerrc`**: in this case, the provider usually states what the update policy will be and you can relax your imports constraints to avoid having to catch up in your `package.json` every time a new version of `frontend-js-react-web$react` is released. This is the way DXP modules build because, traditionally, we haven't had semantic version resolution in DXP (for JavaScript) and because, given that we control the whole product, we want to fully deduplicate and use one exact copy of `react`. So, what we do is placing a `>=n.n.n` constraint on `frontend-js-react-web$react` in `.npmbundlerrc`, so that if the infrastructure is updated to a higher compatible version, the application modules don't need to be rebuilt.
+
+3. **No version constraints in `package.json`, only in `.npmbundlerrc`**: this is a degenerate case of 2 where you simply omit `react` in your build. The benefits are that, if you don't need it for the build, you don't download or care about it. However, it may be considered wrong because you are not declaring one of your dependencies in the `package.json` file even though it will be used during runtime.
+
+> You can see all the implicit imports you get when building a `DXP` module in the [default npm bundler preset](https://github.com/liferay/liferay-npm-tools/blob/master/packages/liferay-npm-bundler-preset-liferay-dev/config.json), that gets injected by `liferay-npm-scripts` when you run `gradle build`. Look for `"imports"` inside the `"config"` section.
+
+> Note that strategy 2 has to be carefully maintained because if you specify a very different version in `package.json` and `.npmbundlerrc` and develop your project targeting the version in `package.json`, you may have problems during runtime if your provider's version is not compatible. This is not the case in `DXP` where we coordinate together to align JavaScript version, but it is worth mentioning the risk.
+
+## How npm bundler imports works at runtime
+
+The technique used to make imports work is quite simple.
```

Comment:
Again, _lies_!

Again, just kidding, but you should delete this kind of stuff that doesn't transmit any information beyond a value judgement.

Reply:
Hey, it's my product. I have to sell it, the same you way you sell `portool`

People has to earn their lives :shrug:

Reply:
Removed for the sake of political correctness :sob:

Reply:
> I have to sell it, the same you way you sell `portool`

You got me there. And I don't even have a single client yet anyway, so I have no credibility.

![despair](https://user-images.githubusercontent.com/7074/92452634-91c66500-f1be-11ea-8ab4-35361f9ab4ad.gif)